common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,1869 @@
+	if err := ms.txStore.UpdateTxsUnconfirmed(ctx, txIDs); err != nil {
+		return err
+	}
+
+	// Update in memory store
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	wg := sync.WaitGroup{}
+	for _, as := range ms.addressStates {
+		wg.Add(1)
+		go func(as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) {
+			for _, txID := range txIDs {
+				if err := as.MoveConfirmedMissingReceiptToUnconfirmed(txID); err != nil {
+					continue
+				}
+			}
+			wg.Done()
+		}(as)
+	}
+	wg.Wait()
+
+	return nil
+}
+
+// FindTxAttemptsRequiringReceiptFetch returns all transactions that are missing a receipt
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTxAttemptsRequiringReceiptFetch(ctx context.Context, chainID CHAIN_ID) (
+	attempts []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	err error,
+) {
+	if ms.chainID.String() != chainID.String() {
+		return attempts, fmt.Errorf("find_tx_attempts_requiring_receipt_fetch: %w", ErrInvalidChainID)
+	}
+
+	filterFn := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+		if tx.TxAttempts != nil && len(tx.TxAttempts) > 0 {
+			attempt := tx.TxAttempts[0]
+			return attempt.State != txmgrtypes.TxAttemptInsufficientFunds
+		}
+
+		return false
+	}
+	states := []txmgrtypes.TxState{TxUnconfirmed, TxConfirmedMissingReceipt}
+	attempts = []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{}
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	for _, as := range ms.addressStates {
+		attempts = append(attempts, as.FetchTxAttempts(states, filterFn)...)
+	}
+	// sort by sequence ASC, gas_price DESC, gas_tip_cap DESC
+	sort.Slice(attempts, func(i, j int) bool {
+		return (*attempts[i].Tx.Sequence).Int64() < (*attempts[j].Tx.Sequence).Int64()
+	})
+
+	// deep copy the attempts
+	var eAttempts []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	for _, attempt := range attempts {
+		eAttempts = append(eAttempts, ms.deepCopyTxAttempt(attempt.Tx, attempt))
+	}
+
+	return eAttempts, nil
+}
+
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTxesPendingCallback(ctx context.Context, blockNum int64, chainID CHAIN_ID) (
+	[]txmgrtypes.ReceiptPlus[R],
+	error,
+) {
+	if ms.chainID.String() != chainID.String() {
+		return nil, fmt.Errorf("find_txes_pending_callback: %w", ErrInvalidChainID)
+	}
+
+	filterFn := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+		if tx.TxAttempts == nil || len(tx.TxAttempts) == 0 {
+			return false
+		}
+
+		if tx.TxAttempts[0].Receipts == nil || len(tx.TxAttempts[0].Receipts) == 0 {
+			return false
+		}
+
+		if tx.PipelineTaskRunID.Valid && tx.SignalCallback && !tx.CallbackCompleted &&
+			tx.TxAttempts[0].Receipts[0].GetBlockNumber() != nil &&
+			big.NewInt(blockNum-int64(tx.MinConfirmations.Uint32)).Cmp(tx.TxAttempts[0].Receipts[0].GetBlockNumber()) > 0 {
```

Comment:
I think we'd also want to allow these to be equal
```suggestion
			big.NewInt(blockNum-int64(tx.MinConfirmations.Uint32)).Cmp(tx.TxAttempts[0].Receipts[0].GetBlockNumber()) >= 0 {
```